File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/organisms/navbar/OrgBottomNavigationBar.kt
package com.perrystreet.woof.designsystem.atomic.organisms.navbar

import androidx.compose.material3.NavigationBar
import androidx.compose.material3.NavigationBarItem
import androidx.compose.material3.NavigationBarItemDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.res.stringResource
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.icon.AtomIcon
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.AtomText
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.atomic.organisms.navbar.roles.BottomNavigationRole
import com.perrystreet.woof.designsystem.theme.Theme

@Immutable
data class OrgBottomNavigationItem(
    val role: BottomNavigationRole,
    val isSelected: Boolean,
    val onTap: () -> Unit,
)

@Composable
fun OrgBottomNavigationBar(items: List<OrgBottomNavigationItem>) {
    NavigationBar(containerColor = Theme.colors.surfaceContainer) {
        items.forEach { item ->
            NavigationBarItem(
                selected = item.isSelected,
                onClick = item.onTap,
                icon = {
                    AtomIcon(
                        iconRes = item.iconRes(),
                        iconSize = SizingRoles.Icon.M,
                        colorRole = item.iconColorRole(),
                    )
                },
                label = {
                    AtomText(
                        text = stringResource(item.role.labelRes),
                        textFontRole = TextFontRole.SubheadP3,
                        colorRole = item.labelColorRole(),
                        maxLines = 1,
                    )
                },
                colors = NavigationBarItemDefaults.colors(indicatorColor = Theme.colors.primary),
            )
        }
    }
}

private fun OrgBottomNavigationItem.iconRes(): Int = when (isSelected) {
    true -> role.selectedIconRes
    false -> role.iconRes
}

private fun OrgBottomNavigationItem.iconColorRole(): IconColorRole = when (isSelected) {
    true -> IconColorRole.OnPrimary
    false -> IconColorRole.OnSurfaceVariant
}

private fun OrgBottomNavigationItem.labelColorRole(): TextColorRole = when (isSelected) {
    true -> TextColorRole.Primary
    false -> TextColorRole.OnSurfaceVariant
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Organisms/NavBar/OrgBottomNavigationBar.swift
import Resources
import SwiftUI

public struct OrgBottomNavigationItem<Value: Hashable>: Identifiable {
    public let icon: ImageAsset
    public let label: String
    public let value: Value
    public let isSelected: Bool
    public let onTap: () -> Void

    public var id: Value { value }

    public init(icon: ImageAsset, label: String, value: Value, isSelected: Bool, onTap: @escaping () -> Void) {
        self.icon = icon
        self.label = label
        self.value = value
        self.isSelected = isSelected
        self.onTap = onTap
    }
}

public struct OrgBottomNavigationBarLabel: View {
    @Environment(\.theme) private var theme

    private let icon: ImageAsset
    private let label: String
    private let isSelected: Bool

    public init(icon: ImageAsset, label: String, isSelected: Bool) {
        self.icon = icon
        self.label = label
        self.isSelected = isSelected
    }

    public var body: some View {
        Label {
            AtomText(text: label, textFontRole: .subheadP3, colorRole: labelColorRole, maxLines: 1)
        } icon: {
            AtomIcon(icon: icon, iconSize: .m, colorRole: iconColorRole)
        }
    }

    private var iconColorRole: IconColorRole {
        switch isSelected {
        case true: .primary
        case false: .onSurfaceVariant
        }
    }

    private var labelColorRole: TextColorRole {
        switch isSelected {
        case true: .primary
        case false: .onSurfaceVariant
        }
    }
}
